{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f7b33030-058c-4d98-b4dd-08ac07e93613",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import decoupler as dc\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import gseapy as gp\n",
    "from gseapy.plot import barplot, dotplot\n",
    "\n",
    "sns.set_context(\"paper\", font_scale=2) # set some nice global plotting settings.\n",
    "\n",
    "from IPython.display import Image # just for embedding plots we want you to try to make without providing code :)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23598f85-b9e4-4425-8661-9f017854fc90",
   "metadata": {},
   "source": [
    "# Saving your processed adata"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1eee0826-1575-4bbd-96cd-69f4a68d5a72",
   "metadata": {},
   "source": [
    "# run in your scanpy notebook after you go through normalization, pca, clustering, umap!\n",
    "\n",
    "# for example:\n",
    "adata.write_h5ad('/data/class/cosmos_cluster6/PUBLIC/ref/example_cortex_adata_processed.h5ad')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88f01843-531d-41b2-9c5c-b737d29649ed",
   "metadata": {},
   "source": [
    "# Load processed adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0a659670-04cb-4e10-b4f6-4282a606d070",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"/data/class/cosmos_cluster6/PUBLIC/ref/example_cortex_adata_processed.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "233884a2-4728-47aa-8275-58487917fdde",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['log1p'][\"base\"] = None # probably have to run this line, known scanpy bug.. https://github.com/scverse/scanpy/issues/2239#issuecomment-1104178881 \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0dfc3d7-cb13-49db-a3c6-04bc60a83c12",
   "metadata": {},
   "source": [
    "# Calculate cluster marker genes using scanpy\n",
    "\n",
    "[Tutorial](https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html#differentially-expressed-genes-as-markers)\n",
    "\n",
    "(You'll probably get a ton of PerformanceWarnings but they're fine to ignore, not errors just warnings about some internal scanpy code) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ebf92239-6e79-49fd-85b1-d0e1e43eb5bd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n"
     ]
    }
   ],
   "source": [
    "# Obtain cluster-specific differentially expressed genes\n",
    "sc.tl.rank_genes_groups(adata, groupby=\"leiden\", method=\"t-test\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c1e866-11af-4b5a-9fae-20110c392dad",
   "metadata": {},
   "source": [
    "## Dot plot of cluster markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0078b99c-9019-411b-8fa7-0e43037f32e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/scanpy/plotting/_dotplot.py:749: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap', 'norm' will be ignored\n",
      "  dot_ax.scatter(x, y, **kwds)\n"
     ]
    }
   ],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata, groupby=\"leiden\", standard_scale=\"var\", n_genes=5\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf099864-744e-43c3-b468-817b3031eaa8",
   "metadata": {},
   "source": [
    "## Example 1: looking further into cluster 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c23601c4-e7f4-425c-a5e5-fa067651705d",
   "metadata": {},
   "source": [
    "### GO term analysis of cluster markers\n",
    "Looking at so many genes is kind of overwhelming, sometimes GO terms help clarify what's going on!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ae725a4-d506-4fbb-9737-42fab582345a",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_cluster0 = sc.get.rank_genes_groups_df(adata, group=\"0\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35db06c3-e4bd-4547-a367-58a21da3ed7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_cluster0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d7017c-b55d-4f32-a725-1bebeb21d41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gp.get_library_name(organism='Mouse') # available gene set libraries -- feel free to uncomment & explore others!!!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5549553c-0455-494d-a6f4-b1a1f4bd5bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_cluster0 = markers_cluster0[(markers_cluster0['pvals_adj'] < 0.01) & (markers_cluster0['logfoldchanges'] > 2)]\n",
    "print(markers_cluster0.shape) # 100-1,000 typically a good number of genes for GO analysis\n",
    "\n",
    "# Perform Reactome Pathway Enrichment Analysis\n",
    "enr_cluster0 = gp.enrichr(gene_list=markers_cluster0['names'], \n",
    "                 organism='Mouse', # change organism to your organism\n",
    "                 gene_sets=['Reactome_2022','GO_Biological_Process_2023'])\n",
    "\n",
    "# Plot results\n",
    "barplot(enr_cluster0.results, cutoff=0.05)\n",
    "dotplot(enr_cluster0.results, cutoff=0.05)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a16b4b6-8136-4d81-858b-09f86f3604d0",
   "metadata": {},
   "source": [
    "#### Check specific genes from GO analysis\n",
    "`enr.results` has more information not shown in plots such as the genes in your list that were in each GO term / reactome pathway."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1d17cc3-8ff5-4408-8d42-fdbd5dd4abf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# can see which genes were in which reactome pathway \n",
    "\n",
    "enr_cluster0.results.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "637687c0-6a43-4df4-b50b-0625caa95fc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# can also see which genes were in which GO term by filtering results\n",
    "\n",
    "enr_cluster0.results[enr_cluster0.results['Gene_set'] == 'GO_Biological_Process_2023'].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac1f5b5-397d-40bd-8547-90f08b0e3fc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "enr_cluster0.results['Genes'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e12c371c-60eb-4117-b771-13ddff1aa2e2",
   "metadata": {},
   "source": [
    "Heatmap-style UMAPs help show if marker expression is shared across clusters, and quickly tell you if expression is more global or specific to a small cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a88ab22-74af-4ff3-a918-c224d4af6172",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.figsize'] = (10, 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "453eb6f4-791a-424c-ac57-a53d7998ea64",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Gria1','Camk2b'], size = 15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab008999-d9b6-4e6b-9674-b39a4f5efad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], size = 15, legend_loc = \"on data\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11dc8de8-1e43-4aaa-b0a7-3046755c5a2b",
   "metadata": {},
   "source": [
    "## Example 2: looking further into cluster 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889ebbfc-8599-4c13-98c0-2bacf306be12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's try a smaller cluster\n",
    "markers_cluster30 = sc.get.rank_genes_groups_df(adata, group=\"30\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b793a535-1d31-4c03-bbcb-5ae108f2c81d",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_cluster30 = markers_cluster30[(markers_cluster30['pvals_adj'] < 0.01) & (markers_cluster30['logfoldchanges'] > 2)]\n",
    "print(markers_cluster30.shape) # 100-1,000 typically a good number of genes for GO analysis\n",
    "\n",
    "# Perform Reactome Pathway Enrichment Analysis\n",
    "enr_cluster30 = gp.enrichr(gene_list=markers_cluster30['names'], \n",
    "                 organism='Mouse', # change organism to your organism\n",
    "                 gene_sets=['Reactome_2022','GO_Biological_Process_2023'])\n",
    "\n",
    "# Plot results\n",
    "barplot(enr_cluster30.results, cutoff=0.05)\n",
    "dotplot(enr_cluster30.results, cutoff=0.05)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ac27983-a229-418f-bb0e-54f700dff9ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "enr_cluster30.results[enr_cluster30.results['Gene_set'] == 'GO_Biological_Process_2023'].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c873b8d-725c-4a85-8841-edfe37fff742",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Odad1','Dnai4','Spag17'], size = 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a22b5e2a-a328-4194-a9a1-d9d8f9da6d31",
   "metadata": {},
   "source": [
    "## Interpreting results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "618dd791-60e8-4956-8765-fd372657f0aa",
   "metadata": {},
   "source": [
    "In our first example (cluster 0), some of the top-weighted GO terms had to do with neuronal systems, synaptic transmissing, etc. Since I know my data is in brain (left cerebral cortex), I can rule out the ones about heart and muscle as being super meaningful -- although of course these tissues are innervated, so those terms are still somewhat helpful! Overall, these terms tell me that the cluster 0 cell type is likely neuronal. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab0cf88d-68b9-476c-93be-230a96636a37",
   "metadata": {},
   "source": [
    "Researching (Googling) marker genes is a tedious task but something that really helps in understanding your tissue and cell types! Also feel free to **ask TAs or Dr. Mortazavi** since we have spent a lot more time researching individual genes and might know something about at least one or two! :) \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb1d9b2-9779-4789-a432-c909ea8964cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Image(filename='results/gria1.png')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bf4668c-b615-4ec6-8893-6a44dff2053a",
   "metadata": {},
   "source": [
    "These results from example #2 in cluster 30 are a little harder to interpret. However, Googling \"cilium assembly brain cell type\" had some interesting results regarding ependymal cells. In addition, other single-cell RNA-seq literature in brain also talks about some of the cluster-specific markers like Spag17. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18641cb9-a642-42d8-ab29-7a55fa708593",
   "metadata": {},
   "outputs": [],
   "source": [
    "Image(filename='results/cilia.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc9721f6-60be-41e2-b3bc-cba264d571a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Image(filename='results/spag17.png')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5865d5b-3957-4541-8a91-84def4bd6738",
   "metadata": {},
   "source": [
    "# Post-annotation pseudobulk differential expression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33126948-a19d-44d2-b8fe-b829f24d09d5",
   "metadata": {},
   "source": [
    "## Create pseudobulk matrix\n",
    "This code assumes your clusters are annotated with celltypes and your adata as a celltype obs column, e.g. `adata.obs['celltype']` exists!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "559871c7-190b-41ed-a60e-15930859155d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata = dc.get_pseudobulk(\n",
    "    adata,\n",
    "    sample_col='Sample',\n",
    "    groups_col='celltype',\n",
    "    obs=adata.obs,\n",
    "    layer='raw_counts',\n",
    "    mode='sum',\n",
    "    min_cells=10,\n",
    "    min_counts=1000\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3632f1-0e48-4c7a-a8d4-1e0e59f4a87e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "751c4d1f-92d3-4b9e-aac6-4cda5f17d8c6",
   "metadata": {},
   "source": [
    "## Compare two groups within each celltype\n",
    "Save results to csv's. Takes a while to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2235a4-37da-45ca-a59d-cc67c7f220e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison = 'Genotype' # change to your comparison of interest!\n",
    "group1 = 'CASTJ' # upregulated genes will be up in this group\n",
    "group2 = 'B6J' # other group for comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "72c39b47-5674-47ea-b32e-fce5c9a91c80",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Neuron\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 21.78 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 26.32 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 15.67 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 4.38 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.52 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>3.210471</td>\n",
       "      <td>0.362502</td>\n",
       "      <td>0.738919</td>\n",
       "      <td>0.490584</td>\n",
       "      <td>0.623721</td>\n",
       "      <td>0.775332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.417554</td>\n",
       "      <td>0.967909</td>\n",
       "      <td>1.805823</td>\n",
       "      <td>0.535993</td>\n",
       "      <td>0.591963</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>7.032265</td>\n",
       "      <td>-0.107681</td>\n",
       "      <td>0.540918</td>\n",
       "      <td>-0.199072</td>\n",
       "      <td>0.842207</td>\n",
       "      <td>0.917393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>45.031372</td>\n",
       "      <td>0.005866</td>\n",
       "      <td>0.149217</td>\n",
       "      <td>0.039310</td>\n",
       "      <td>0.968643</td>\n",
       "      <td>0.985624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>4.149971</td>\n",
       "      <td>0.138672</td>\n",
       "      <td>0.487919</td>\n",
       "      <td>0.284210</td>\n",
       "      <td>0.776249</td>\n",
       "      <td>0.878683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.226185</td>\n",
       "      <td>1.006925</td>\n",
       "      <td>3.816373</td>\n",
       "      <td>0.263843</td>\n",
       "      <td>0.791901</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.099974</td>\n",
       "      <td>-0.435760</td>\n",
       "      <td>3.872784</td>\n",
       "      <td>-0.112519</td>\n",
       "      <td>0.910412</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.099974</td>\n",
       "      <td>-0.435760</td>\n",
       "      <td>3.872784</td>\n",
       "      <td>-0.112519</td>\n",
       "      <td>0.910412</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.140737</td>\n",
       "      <td>1.006924</td>\n",
       "      <td>3.873269</td>\n",
       "      <td>0.259968</td>\n",
       "      <td>0.794889</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.100298</td>\n",
       "      <td>-0.435760</td>\n",
       "      <td>3.872784</td>\n",
       "      <td>-0.112519</td>\n",
       "      <td>0.910412</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "gene_name                                                                \n",
       "0610005C13Rik   3.210471        0.362502  0.738919  0.490584  0.623721   \n",
       "0610006L08Rik   0.417554        0.967909  1.805823  0.535993  0.591963   \n",
       "0610009B22Rik   7.032265       -0.107681  0.540918 -0.199072  0.842207   \n",
       "0610009E02Rik  45.031372        0.005866  0.149217  0.039310  0.968643   \n",
       "0610009L18Rik   4.149971        0.138672  0.487919  0.284210  0.776249   \n",
       "...                  ...             ...       ...       ...       ...   \n",
       "mt-Ti           0.226185        1.006925  3.816373  0.263843  0.791901   \n",
       "mt-Tl1          0.099974       -0.435760  3.872784 -0.112519  0.910412   \n",
       "mt-Tm           0.099974       -0.435760  3.872784 -0.112519  0.910412   \n",
       "mt-Tr           0.140737        1.006924  3.873269  0.259968  0.794889   \n",
       "mt-Tv           0.100298       -0.435760  3.872784 -0.112519  0.910412   \n",
       "\n",
       "                   padj  \n",
       "gene_name                \n",
       "0610005C13Rik  0.775332  \n",
       "0610006L08Rik       NaN  \n",
       "0610009B22Rik  0.917393  \n",
       "0610009E02Rik  0.985624  \n",
       "0610009L18Rik  0.878683  \n",
       "...                 ...  \n",
       "mt-Ti               NaN  \n",
       "mt-Tl1              NaN  \n",
       "mt-Tm               NaN  \n",
       "mt-Tr               NaN  \n",
       "mt-Tv               NaN  \n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/pandas/core/arraylike.py:396: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oligodendrocyte\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 11.45 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 15.50 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 6.95 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.50 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.50 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.080965</td>\n",
       "      <td>0.116604</td>\n",
       "      <td>3.870039</td>\n",
       "      <td>0.030130</td>\n",
       "      <td>0.975964</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.121242</td>\n",
       "      <td>1.559288</td>\n",
       "      <td>3.870118</td>\n",
       "      <td>0.402904</td>\n",
       "      <td>0.687019</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.749481</td>\n",
       "      <td>-0.572073</td>\n",
       "      <td>1.633619</td>\n",
       "      <td>-0.350188</td>\n",
       "      <td>0.726198</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE      stat    pvalue  padj\n",
       "gene_name                                                                  \n",
       "0610005C13Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610006L08Rik  0.080965        0.116604  3.870039  0.030130  0.975964   NaN\n",
       "0610009B22Rik  0.121242        1.559288  3.870118  0.402904  0.687019   NaN\n",
       "0610009E02Rik  0.749481       -0.572073  1.633619 -0.350188  0.726198   NaN\n",
       "0610009L18Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "...                 ...             ...       ...       ...       ...   ...\n",
       "mt-Ti          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tl1         0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tm          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tr          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tv          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Endothelial\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/pandas/core/arraylike.py:396: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 7.12 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 10.09 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 3.41 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.45 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.51 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.246767</td>\n",
       "      <td>-1.739439</td>\n",
       "      <td>4.13126</td>\n",
       "      <td>-0.421043</td>\n",
       "      <td>0.673724</td>\n",
       "      <td>0.985011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange    lfcSE      stat    pvalue      padj\n",
       "gene_name                                                                     \n",
       "0610005C13Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "0610006L08Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "0610009B22Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "0610009E02Rik  0.246767       -1.739439  4.13126 -0.421043  0.673724  0.985011\n",
       "0610009L18Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "...                 ...             ...      ...       ...       ...       ...\n",
       "mt-Ti          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tl1         0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tm          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tr          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tv          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Astrocyte\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 13.25 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 17.66 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 7.95 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.99 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.70 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.147358</td>\n",
       "      <td>1.043214</td>\n",
       "      <td>3.868723</td>\n",
       "      <td>0.269653</td>\n",
       "      <td>0.787427</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.345654</td>\n",
       "      <td>0.987975</td>\n",
       "      <td>2.418431</td>\n",
       "      <td>0.408519</td>\n",
       "      <td>0.682893</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>1.814146</td>\n",
       "      <td>1.620366</td>\n",
       "      <td>1.100884</td>\n",
       "      <td>1.471878</td>\n",
       "      <td>0.141054</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>1.007605</td>\n",
       "      <td>0.299025</td>\n",
       "      <td>1.346066</td>\n",
       "      <td>0.222147</td>\n",
       "      <td>0.824199</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE      stat    pvalue  padj\n",
       "gene_name                                                                  \n",
       "0610005C13Rik  0.147358        1.043214  3.868723  0.269653  0.787427   NaN\n",
       "0610006L08Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610009B22Rik  0.345654        0.987975  2.418431  0.408519  0.682893   NaN\n",
       "0610009E02Rik  1.814146        1.620366  1.100884  1.471878  0.141054   NaN\n",
       "0610009L18Rik  1.007605        0.299025  1.346066  0.222147  0.824199   NaN\n",
       "...                 ...             ...       ...       ...       ...   ...\n",
       "mt-Ti          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tl1         0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tm          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tr          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tv          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Microglia\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 9.02 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 13.11 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/pydeseq2/dds.py:558: RuntimeWarning: invalid value encountered in log\n",
      "  ) - np.log(self[:, self.non_zero_genes].varm[\"fitted_dispersions\"])\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/numpy/lib/utils.py:1119: ImplicitModificationWarning: Trying to modify attribute `.varm` of view, initializing view as actual.\n",
      "  result[n] = np.nan\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... done in 5.08 seconds.\n",
      "\n",
      "Fitting LFCs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/pydeseq2/dds.py:623: RuntimeWarning: invalid value encountered in log\n",
      "  self.varm[\"_outlier_genes\"] = np.log(self.varm[\"genewise_dispersions\"]) > np.log(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... done in 2.39 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.52 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.156138</td>\n",
       "      <td>-0.695461</td>\n",
       "      <td>3.798462</td>\n",
       "      <td>-0.183090</td>\n",
       "      <td>0.854727</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0.078069</td>\n",
       "      <td>-0.089200</td>\n",
       "      <td>3.875363</td>\n",
       "      <td>-0.023017</td>\n",
       "      <td>0.981637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE      stat    pvalue  padj\n",
       "gene_name                                                                  \n",
       "0610005C13Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610006L08Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610009B22Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610009E02Rik  0.156138       -0.695461  3.798462 -0.183090  0.854727   NaN\n",
       "0610009L18Rik  0.078069       -0.089200  3.875363 -0.023017  0.981637   NaN\n",
       "...                 ...             ...       ...       ...       ...   ...\n",
       "mt-Ti          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tl1         0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tm          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tr          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tv          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OPC\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 10.56 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 14.59 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 6.21 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.59 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.72 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.483216</td>\n",
       "      <td>1.390946</td>\n",
       "      <td>2.229641</td>\n",
       "      <td>0.623843</td>\n",
       "      <td>0.532731</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0.234838</td>\n",
       "      <td>0.189070</td>\n",
       "      <td>3.051911</td>\n",
       "      <td>0.061951</td>\n",
       "      <td>0.950602</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE      stat    pvalue  padj\n",
       "gene_name                                                                  \n",
       "0610005C13Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610006L08Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610009B22Rik  0.483216        1.390946  2.229641  0.623843  0.532731   NaN\n",
       "0610009E02Rik  0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "0610009L18Rik  0.234838        0.189070  3.051911  0.061951  0.950602   NaN\n",
       "...                 ...             ...       ...       ...       ...   ...\n",
       "mt-Ti          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tl1         0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tm          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tr          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "mt-Tv          0.000000             NaN       NaN       NaN       NaN   NaN\n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VLMC\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 8.02 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 11.33 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 3.78 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.35 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 4.51 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: Genotype CASTJ vs B6J\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0.133707</td>\n",
       "      <td>-0.694583</td>\n",
       "      <td>4.42567</td>\n",
       "      <td>-0.156944</td>\n",
       "      <td>0.875289</td>\n",
       "      <td>0.984037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Ti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tl1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tr</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tv</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange    lfcSE      stat    pvalue      padj\n",
       "gene_name                                                                     \n",
       "0610005C13Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "0610006L08Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "0610009B22Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "0610009E02Rik  0.133707       -0.694583  4.42567 -0.156944  0.875289  0.984037\n",
       "0610009L18Rik  0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "...                 ...             ...      ...       ...       ...       ...\n",
       "mt-Ti          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tl1         0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tm          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tr          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "mt-Tv          0.000000             NaN      NaN       NaN       NaN       NaN\n",
       "\n",
       "[29897 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pericyte\n",
      "Skipping Pericyte because both groups are not present.\n",
      "Epithelial\n",
      "Skipping Epithelial because both groups are not present.\n",
      "Fibroblast\n",
      "Skipping Fibroblast because both groups are not present.\n",
      "Ependymal\n",
      "Skipping Ependymal because both groups are not present.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/opt/apps/python/3.8.0/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "types = adata.obs['celltype'].unique()\n",
    "\n",
    "for this_type in types:\n",
    "    print(this_type)\n",
    "    gdata = pdata[pdata.obs[comparison].isin([group1, group2])].copy() # filter for the 2 groups of interest\n",
    "    ctdata = gdata[gdata.obs['celltype'] == this_type].copy()\n",
    "\n",
    "    # Count the number of samples in each group\n",
    "    group_counts = ctdata.obs[comparison].value_counts()\n",
    "    \n",
    "    # Make sure each group is present\n",
    "    if (group_counts.get(group1, 0) >= 1) and (group_counts.get(group2, 0) >= 1):\n",
    "        dds = DeseqDataSet(adata=ctdata,\n",
    "                           design_factors=comparison, \n",
    "                           refit_cooks=True)\n",
    "        \n",
    "        dds.deseq2()\n",
    "        stat_res = DeseqStats(dds, contrast=[comparison, group1, group2])\n",
    "        stat_res.summary()\n",
    "        \n",
    "        results = stat_res.results_df # get results df\n",
    "        \n",
    "        results['celltype'] = this_type\n",
    "        results['nlog10padj'] = -np.log10(results['padj']) # make -log10(adj. p-value) column\n",
    "        results.replace([np.inf, -np.inf], np.nan, inplace=True) # remove NAN and infinity values...\n",
    "        results = results.dropna() # drop NAs\n",
    "        results = results.reset_index().rename(columns={'index': 'gene_name'}) # set gene name to be its own column\n",
    "        results.to_csv(f'results/deg_results_{this_type}_{group1}_vs_{group2}.csv', index=False)\n",
    "    else:\n",
    "        print(f\"Skipping {this_type} because both groups are not present.\")\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9b40f09-af35-4d47-8a27-d7e3c9a094f6",
   "metadata": {},
   "source": [
    "## Dot plot of DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b34543e8-a710-4c4b-9991-91afbf505370",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = \"Oligodendrocyte_CASTJ_vs_B6J\" # example!! change for your file!\n",
    "results = pd.read_csv(f\"results/deg_results_{fname}.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c859d9-e704-4385-80d5-21164eedaca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# at celltype level -- maybe not too interesting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48cf433d-b8b0-4a5c-963a-d27f96ac1145",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs_up = results[(results['padj'] < 0.05) & (results['log2FoldChange'] > 2)] # increased lfc threshold for less genes to show\n",
    "degs_down = results[(results['padj'] < 0.05) & (results['log2FoldChange'] < -2)] \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf033d3a-d37b-41b1-a3de-5b34bb054dd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, degs_up['gene_name'].tolist(), 'celltype', \n",
    "              mean_only_expressed = False,\n",
    "              dendrogram=True, log=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bf71ef5-55e2-47f3-a125-6c0a5ade8612",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, degs_down['gene_name'].tolist(), 'celltype', \n",
    "              mean_only_expressed = False,\n",
    "              dendrogram=True, log=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37d07cd0-b42d-4397-938a-264b2767eb0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try filtering the adata for the celltype of interest and grouping by comparison\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5603c6f4-1f1e-4119-aa58-8740435f705f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_thistype = adata[adata.obs['celltype'] == 'Oligodendrocyte'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b393fdb-33e5-404e-afe4-af7594ded75f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata_thistype, degs_up['gene_name'].tolist(), 'Genotype', \n",
    "              mean_only_expressed = False,\n",
    "              dendrogram=True, log=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49355862-fca3-4fda-91b5-f1c6e2b7f105",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata_thistype, degs_down['gene_name'].tolist(), 'Genotype', \n",
    "              mean_only_expressed = False,\n",
    "              dendrogram=True, log=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ecddcb2-7e17-4019-b5c8-768f620890ff",
   "metadata": {},
   "source": [
    "## Volcano plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df045881-da60-409d-a5b9-96d27bf2b0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "! pip install adjustText # in case u dont have this package\n",
    "\n",
    "from adjustText import adjust_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d6e086-51dd-4951-b25a-97ee4dcf2cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = \"Oligodendrocyte_CASTJ_vs_B6J\" # example!! change for your file!\n",
    "results = pd.read_csv(f\"results/deg_results_{fname}.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "922816eb-3fdd-4848-acde-8f82843ef510",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs_up = results[(results['padj'] < 0.05) & (results['log2FoldChange'] > 1)]\n",
    "degs_down = results[(results['padj'] < 0.05) & (results['log2FoldChange'] < -1)]\n",
    "\n",
    "\n",
    "plt.figure(figsize=(7, 6))\n",
    "sns.scatterplot(data=results, x='log2FoldChange', y='nlog10padj', alpha=0.5, \n",
    "                color='black', label = \"Not significant\")\n",
    "\n",
    "# basically layer the subset dataframes on top of the original plot\n",
    "sns.scatterplot(data=degs_up, x='log2FoldChange', y='nlog10padj', alpha=0.5, \n",
    "                color='red', label = f\"Upregulated ({len(degs_up)} genes)\")\n",
    "sns.scatterplot(data=degs_down, x='log2FoldChange', y='nlog10padj', alpha=0.5, \n",
    "                color='blue', label = f\"Downregulated ({len(degs_down)} genes)\")\n",
    "\n",
    "plt.xlabel('log2(Fold Change)')\n",
    "plt.ylabel('-log10(adj. p-value)')\n",
    "plt.title(fname)\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc='upper left') # move legend\n",
    "\n",
    "# draw horizontal and vertical dashed lines\n",
    "plt.axhline(1.3, color='black', linestyle='--')\n",
    "plt.axvline(1, color='black', linestyle='--')\n",
    "plt.axvline(-1, color='black', linestyle='--')\n",
    "\n",
    "# sry this should help fix the overlaps\n",
    "texts = []\n",
    "for index, row in results.iterrows():\n",
    "    if abs(row['log2FoldChange']) > 1 and row['nlog10padj'] > 10:\n",
    "        texts.append(plt.text(row['log2FoldChange'], row['nlog10padj'], row['gene_name'], fontsize=10))\n",
    "\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle='-', color='black'))\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "678bb1be-6364-458a-a5b4-9a2a84916f94",
   "metadata": {},
   "source": [
    "## GO term enrichment of DEGs\n",
    "\n",
    "Up- and down-regulated **separately**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ac602e6-88c4-4408-98e6-9b8302c59877",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs_up = results[(results['padj'] < 0.05) & (results['log2FoldChange'] > 0.2)] # not many DEGs in my case so had to decrease LFC threshold...\n",
    "\n",
    "enr_up = gp.enrichr(gene_list=degs_up['gene_name'], \n",
    "                 organism='Mouse', # change organism to your organism\n",
    "                 gene_sets=['Reactome_2022','GO_Biological_Process_2023'])\n",
    "\n",
    "barplot(enr_up.results, cutoff=0.05)\n",
    "dotplot(enr_up.results, cutoff=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e042590d-682f-4785-8125-8a8bffcd5bc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs_down = results[(results['padj'] < 0.05) & (results['log2FoldChange'] < -1)]\n",
    "\n",
    "enr_down = gp.enrichr(gene_list=degs_down['gene_name'], \n",
    "                 organism='Mouse', # change organism to your organism\n",
    "                 gene_sets=['Reactome_2022','GO_Biological_Process_2023'])\n",
    "\n",
    "barplot(enr_down.results, cutoff=0.05)\n",
    "dotplot(enr_down.results, cutoff=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac991122-a551-4da9-957d-98807845eff0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32e8b16-62dd-4e34-9f61-f118d0768b1d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
